### PR TITLE
Expose DHT content providers API from `sc-network`

### PR DESCRIPTION
**Pull request by dmitry-markin**
Expose the Kademlia content providers API for the use by `sc-network` client code:
1. Extend the `NetworkDHTProvider` trait with functions to start/stop providing content and query the DHT for the list of content providers for a given key.
2. Extend the `DhtEvent` enum with events reporting the found providers or query failures.
3. Implement the above for libp2p & litep2p network backends.

### CODE REVIEW

**Comment by dmitry-markin** [comment]
It looks like the CI always gets stuck after a `/cmd prdoc` command. Another PR where I had to manually kick the CI after `/cmd prdoc` is https://github.com/paritytech/polkadot-sdk/pull/6703.

CC @alvicsam

File: substrate/client/network/src/litep2p/mod.rs [hunk 1]
```diff
@@ -163,6 +163,9 @@ pub struct Litep2pNetworkBackend {
 	/// Pending `PUT_VALUE` queries.
 	pending_put_values: HashMap<QueryId, (RecordKey, Instant)>,
 
+	/// Pending `GET_PROVIDERS` queries.
+	pending_get_providers: HashMap<QueryId, (RecordKey, Instant)>,
```
Comment (lexnv) [inline substrate/client/network/src/litep2p/mod.rs:167]: nit: We could merge `pending_get_values`, `pending_put_values` and `pending_get_providers` into a single `pending_queries`.

The new `pending_queries` can hold sufficient information to distinguish between query types:

```rust
pending_queries: HashMap<QueryId, QueryType>

enum QueryType {
    GetValue((RecordKey, Instant)),
    PutValue((RecordKey, Instant)),
    GetProviders((RecordKey, Instant)),
}
```

The advantage of doing so is that we would remove duplicated code that handles the case where the queryID is not found in our mappings. At the same time, we will never forget about checking a hashmap basically turning code like:

```
if let Some(..) = pending_get_values.get(..) {

} else if let Some(..) = pending_put_values.get(..) {

} else if let Some(..) = pending_queries.get(..) {
    ..
}

```
Into a more easier to follow:

```
match pending_queries.get(..) {
    None => // absent,
    Some(GetValue(..))
    ...
}
```

Up to you in the end

File: substrate/client/network/src/discovery.rs [hunk 2]
```diff
@@ -581,6 +606,15 @@ pub enum DiscoveryOut {
 	/// Returning the corresponding key as well as the request duration.
 	ValuePutFailed(RecordKey, Duration),
 
+	/// Starting providing a key failed.
+	StartProvidingFailed(RecordKey),
```
Comment (lexnv) [inline substrate/client/network/src/discovery.rs:610]: Would passing the `Duration` be difficult?

File: substrate/client/network/src/discovery.rs [hunk 3]
```diff
@@ -982,6 +1016,56 @@ impl NetworkBehaviour for DiscoveryBehaviour {
...
+						result: QueryResult::GetProviders(res),
+						stats,
+						id,
+						..
+					} => {
+						let ev = match res {
+							Ok(GetProvidersOk::FoundProviders { key, providers }) => {
+								debug!(
+									target: "sub-libp2p",
+									"Libp2p => Found providers {:?} for key {:?},  id {:?}, stats {:?}",
```
Comment (lexnv) [inline substrate/client/network/src/discovery.rs:1029]: nit: 
```suggestion
									"Libp2p => Found providers {:?} for key {:?}, id {:?}, stats {:?}",
```

**Review by lexnv — APPROVED** [review]
LGTM!

**Review by bkchr — APPROVED** [review]

**Pull request merged**